src/sys/petsclib.cxx

```diff
@@ -22,6 +22,22 @@ PetscLib::PetscLib() {
     PetscInitialize(pargc,pargv,PETSC_NULL,help);
     PetscLogEventRegister("Total BOUT++",0,&USER_EVENT);
     PetscLogEventBegin(USER_EVENT,0,0,0,0);
+
+    // Load global PETSc options from the [petsc] section of the input
+    setPetscOptions(Options::root()["petsc"], "");
+  }
+
+  if (opt != nullptr and opt->isSection()) {
+    // Use options specific to this PetscLib
+    // Pass options to PETSc's global options database, with a unique prefix, that will be
+    // passed to a KSP later.
+    // (PetscOptions type exists for non-global options, but apparently is only for user
+    // options, and cannot be passed to KSP, etc. Non-global options can be passed by
+    // defining a custom prefix for the options string, and then passing that to the KSP.)
+
+    options_prefix = "boutpetsclib" + std::to_string(count) + "_";
```

Comment:
I suspect this could be non-unique. The counter gets incremented on construction and decremented on destruction so we could have something like

`
object someParentObject;
{
auto libA = make_a_lib(); //Make a lib instance in this scope  -- count is 1
SomeParentObject.lib = make_a_lib(); // Make a lib instance for something else -- count is 2
} // Leave the scope so libA is destroyed, count is now 1
auto libB = make_a_lib(); // count is 2 again

`

Reply:
Good spot! I'd hope unlikely in practice, but it would be a pig to find if it ever happened. Would the easiest fix would be to just add another counter that only goes up?

Reply:
Actually I've just realised `count` is of course static so has a single value across all instances at the time of access. The above concern still applies but there's a further point that we only care about the value of count at the point of creation as this is when the options will be registered, but therefore I don't think we should need to keep the instance of the lib as this really just holds the unique `options_prefix`, so we can instead just hold the `ksp` instance that has this prefix set _or_ hold the prefix and pass that to the ksp creation routine.

Reply:
Is keeping instances of `PetscLib` a problem? It looks like the objects are very lightweight - basically just reference counting so that `PetscInitialize()` is called when the first `PetscLib` is created and `PetscFinalize()` is called when the last one is destroyed. If that's right, maybe the name is a bit misleading... It seems like a reasonable place to me to also hide away the `options_prefix` from the solvers that use PETSc, I think it probably simplifies the solver code a bit not to have to handle the prefix explicitly.